<img src="figs/cit_hub_logo.png" alt="logo" width="300"/>


# Multiple Conditions

So far, we've only worked with single conditions, like the following:

- If `price` equals `0.0` (`if price == 0`)
- If `genre` equals `"Games"` (`if genre == 'Games'`)

Single conditions won't allow us to answer more advanced questions, like these:

- What's the average rating of free gaming apps?
- What's the average rating of not-free gaming apps?
- What's the average rating of free non-gaming apps?
- What's the average rating of not-free non-gaming apps?

Fortunately, we can combine two or more conditions together into a single if statement using the and keyword. In the two diagrams below, we can use and to simultaneously check whether an app is both free and has a gaming genre.

In [2]:
app_1_price = 1
app_1_genre = 'Games'

if app_1_price == 0 and app_1_genre == 'Games':
    print('This is a free game!')

print(app_1_price == 0 and app_1_genre == 'Games')

False


In [2]:
app_2_price = 19
app_2_genre = 'Games'

if app_2_price == 0 and app_2_genre == 'Games':
    print('This is a free game!')

print(app_2_price == 0 and app_2_genre == 'Games')

False


Notice above that code like `app_1_price == 0 and app_1_genre == 'Games'` outputs a single Boolean value.

![lists](figs/17-if.svg)

Python evaluates any combination of Booleans to a single Boolean value:



In [4]:
print(True and True)
print(True and False)
print(False and True)
print(False and False)

True
False
False
False


As a rule, when we combine Booleans using and **the resulting Boolean is `True` *only if* all the Booleans are `True`**. If any of the Booleans are `False`, then the resulting Boolean will be `False`:

In [5]:
print(True and True and True)
print(True and True and True and True)
print(True and True and False)
print(True and True and True and False)

True
True
False
False


### Determine the average rating of free gaming apps

Now let's use multiple conditions to determine the average rating of free gaming apps.

In [4]:
opened_file = open('AppleStore.csv', encoding="utf")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

free_games_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    genre = row[11]
    # Complete code from here
    if price == 0.0 and genre == 'Games':
        free_games_ratings.append(rating)
        
avg_rating_free_games = sum(free_games_ratings) / len(free_games_ratings)
print(avg_rating_free_games)

3.5285777580859548


If we look at the first five apps, we can see in the prime_genre column that Facebook's genre is "Social Networking," while the genre for Clash of Clans and Temple Run is "Games":

Since social networking apps and games are usually popular, we may want to further investigate this category. One thing we might want to determine is the average rating of this category that encompasses games and social networking apps.

To do that, we need to isolate the ratings of all the apps whose genre is either "Social Networking" or "Games" into a separate list. Then, we can compute the average value using techniques we already know.

If we want to isolate the ratings of these apps using the condition `if genre == 'Social Networking' and genre == 'Games'`, we'll end up with an empty list because there's no app with a genre that is both "Social Networking" and "Games" — an app can only have one genre.

In [6]:
games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]

    if genre == 'Social Networking' or genre == 'Games':
        games_social_ratings.append(rating)

print(games_social_ratings)
print(len(games_social_ratings))

[4.0, 4.0, 4.0, 4.5, 3.5, 4.5, 3.5, 4.0, 5.0, 3.0, 3.5, 3.5, 4.0, 3.0, 4.0, 3.5, 4.0, 4.5, 4.0, 4.5, 2.0, 4.5, 4.0, 3.5, 4.0, 3.0, 4.0, 4.0, 3.5, 4.5, 3.5, 4.5, 4.0, 3.5, 4.0, 4.5, 4.0, 4.5, 4.5, 3.5, 4.5, 4.0, 3.5, 4.0, 4.0, 4.0, 4.0, 3.5, 4.0, 3.5, 2.5, 3.5, 4.0, 4.5, 3.5, 4.0, 3.5, 4.0, 4.0, 3.5, 4.0, 4.0, 4.5, 4.5, 4.0, 4.0, 4.0, 4.0, 5.0, 4.0, 4.0, 3.0, 3.5, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.5, 4.5, 4.0, 3.5, 3.5, 4.0, 4.0, 4.0, 3.0, 5.0, 3.5, 4.5, 4.5, 4.5, 4.5, 3.5, 3.0, 4.0, 4.0, 4.0, 4.5, 4.0, 4.5, 4.0, 4.5, 5.0, 3.5, 3.0, 3.5, 4.0, 3.5, 4.0, 4.5, 3.0, 4.5, 4.0, 3.5, 1.5, 4.5, 5.0, 4.5, 4.0, 4.5, 4.5, 4.0, 3.5, 4.0, 4.0, 4.5, 4.5, 4.0, 4.0, 4.0, 3.5, 3.0, 4.5, 4.0, 4.0, 4.5, 4.5, 4.0, 4.5, 4.0, 4.5, 4.0, 3.5, 5.0, 3.5, 4.5, 3.5, 3.5, 3.5, 3.5, 4.0, 3.5, 4.5, 4.5, 4.5, 3.0, 4.0, 4.5, 4.5, 4.5, 0.0, 4.0, 4.5, 4.5, 4.5, 4.0, 4.5, 4.5, 4.5, 4.5, 0.0, 4.0, 4.5, 3.5, 3.5, 3.5, 4.5, 4.5, 4.5, 4.0, 5.0, 4.5, 4.5, 4.0, 4.5, 3.5, 4.5, 4.0, 3.5, 4.0, 4.0, 4.0, 4.5, 4.5, 4.5, 4.5, 4.0,

We need to isolate the rating of an app only if the genre is "Social Networking" **or** "Games," not "Social Networking" and "Games." To account for this difference, we can use `or` instead of `and`:

In [10]:
games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]

    if genre == 'Social Networking' or genre == 'Games':
        games_social_ratings.append(rating)

print(games_social_ratings[:5])
print(len(games_social_ratings))

[4.0, 4.0, 4.0, 4.5, 3.5]
4029


We call or and and logical operators. Just like and, or unites two or more Booleans. As we learned previously, Python evaluates any combination of Booleans to a single Boolean value:

In [11]:
print(True or True)
print(True or False)
print(False or True)
print(False or False)

True
True
True
False


When we combine Booleans using or, **the resulting Boolean is `False` *only if* all the Booleans are `False`**. If any of the Booleans are `True`, then the resulting Boolean is `True`:

In [12]:
print(False or False or False)
print(False or False or False or True)

False
True


Returning to our apps example, the condition `if genre == 'Social Networking' or genre == 'Games'` resolves to `False` when an app's genre is neither "Social Networking" nor "Games." Otherwise, it resolves to `True`.

### Average rating of apps belonging to a specific genre

Now let's practice using the or operator by computing the average rating of the apps with a genre of either "Social Networking" or "Games."

In [14]:
opened_file = open('AppleStore.csv', encoding="utf-8")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

games_social_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    # Complete code from here
    
    if genre == 'Social Networking' or genre == 'Games':
        games_social_ratings.append(rating)
        
avg_games_social = sum(games_social_ratings) / len(games_social_ratings)
print(avg_games_social)

3.655994043186895


In the previous exercise, we computed the average rating of the apps with a genre that is either "Social Networking" or "Games." We can ask even more specific questions, like these:

- What is the average rating of free apps with a genre that is either "Social Networking" or "Games"?
- What is the average rating of not free apps with a genre that is either "Social Networking" or "Games"?

To answer the first question, we need to isolate the apps that meet both criteria:

- Are in either the "Social Networking" or "Games" genre
- Have a price of 0.0

To isolate these apps, we can combine `or` with `and` in a single `if` statement:



In [21]:
# We want this logic:
if (genre == 'Social Networking' or genre == 'Games') and price ==0;

# Not this logic:
if genre == 'Social Networking' or (genre == 'Games' and price ==0);

SyntaxError: invalid syntax (<ipython-input-21-01c216873f14>, line 2)

If we don't use the parentheses, Python defaults to unwanted results — like incorrectly including not-free apps. In the example below, we see that not using parentheses leads us to classifying a not-free app as free.

![lists](figs/18-if.svg)

Above, we printed 'This gaming or social networking app is free!!' for a not-free app. This is an example of a **logical error**.

In the case of a syntax or runtime error, the computer stops the code from running and raises an error. For logical errors, the computer doesn't raise any error — the code runs just fine and the logical error is unnoticed. To prevent this, we should pay extra attention to the logic of our code.

Using parentheses correctly makes a difference:

In [22]:
app_genre = 'Social Networking'
app_price = 100

print((True or False) and False)

if (app_genre == 'Social Networking' or app_genre == 'Games') and app_price == 0:
    print('This gaming or social networking app is free!')

False


### Average rating of not-free apps of a particular genre

Now, we compute the average rating of not-free apps with a genre that is either "Social Networking" or "Games."

In [24]:
opened_file = open('AppleStore.csv', encoding="utf-8")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

not_free_games_social_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    if (genre == 'Social Networking' or genre == 'Games') and price != 0:
        not_free_games_social_ratings.append(rating)
        
avg_not_free = sum(not_free_games_social_ratings) / len(not_free_games_social_ratings)
print(avg_not_free)

3.8904235727440146


### Boolean Operators in Python

Previously, we used the `==` and `!=` operators to check if two values are equal. When we check for equality, we compare one value to another to determine if they are equal. For this reason, we call `==` and `!=` **comparison operators**.

We can compare value A to value B to determine the following:

- A is equal to B and vice versa (B is equal to A).
- A is not equal to B and vice versa.
- A is greater than B or vice versa.
- A is greater than or equal to B or vice versa.
- A is less than B or vice versa.
- A is less than or equal to B or vice versa.

In Python, we have special operators for each of the comparison operations above:

![lists](figs/19-compare.svg)

As with `==` and `!=`, comparing values using any of the comparison operators above will output a single Boolean value:


In [25]:
print(8 > 1)
print(8 < 1)
print(30 >= 30)
print(20 <= 19)

True
False
True
False


### Comparison operators

Comparison operators allow us to answer more advanced questions about our dataset:

- How many apps have a rating of `4.0` or greater?
- What is the average rating of the apps that have a price greater than `$9`?
- How many apps have a price greater than `$9`?
- How many apps have a price less than or equal to `$9`?

Using what we know, we can answer the first question in at least two ways.

In [26]:
apps_4_or_greater = []

for row in apps_data[1:]:
    rating = float(row[7])
    if rating >= 4.0:
        apps_4_or_greater.append(rating)

print(len(apps_4_or_greater))

4781


Above, we did the following:

- Initialized an empty list named `apps_4_or_greater`.
- Looped through `apps_data[1:]`, and for every iteration, we:
    - Stored the rating value as a float to a variable named `rating`.
    - Appended the rating to `apps_4_or_greater` if the value stored in rating was greater than or equal to 4.0.
- Measured the length of the `apps_4_or_greater` list to find out the number of apps that have a rating of 4.0 or greater.
    - After the loop, the `apps_4_or_greater` list only stores the ratings of the apps that were rated 4.0 or better. The list `apps_4_or_greater` has 4,781 ratings that are 4.0 or greater, which means that there are 4,781 apps that have a rating of 4.0 or greater.

We can also use another approach to answer the question: we initialize a variable with a value of 0 and then increment that variable by 1 each time we find a rating of 4.0 or greater:

In [27]:
n_of_apps = 0

for row in apps_data[1:]:
    rating = float(row[7])
    if rating >= 4.0:
        n_of_apps = n_of_apps + 1

print(n_of_apps)

4781


### Further explore our dataset

Now let's answer the other three questions:

- What is the average rating of the apps that have a price greater than `$9`?
- How many apps have a price greater than `$9`?
- How many apps have a price less than or equal to `$9`?

In [30]:
opened_file = open('AppleStore.csv', encoding="utf-8")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    
    if price > 9:
        ratings.append(rating)
        
avg_rating = sum(ratings) / len(ratings)
n_apps_more_9 = len(ratings)
n_apps_less_9 = len(apps_data[1:]) - len(ratings)

print(f"Average rating is: {avg_rating}")
print(f"There are {n_apps_more_9} apps with a price greater than 9$")
print(f"There are {n_apps_less_9} apps with a price less than or equal to 9$")

Average rating is: 3.5280898876404496
There are 178 apps with a price greater than 9$
There are 7019 apps with a price less than or equal to 9$


Let's use information from the price column to label each app as "`free`" or "`not free`." If the price equals 0.0, we want to label the app "free." Otherwise, we want to label it "not free."

Remember that we store our dataset as a list of lists — each row is a list that describes an app. To label the app, we want to add the string 'free' or 'not free' at the end of each list (row). On a smaller scale, this is what we want to do (below, we're using just a small extract from our dataset, showing just the name and the price of four apps):

In [31]:
apps_data = [['Call of Duty: Zombies', 5.0], ['Facebook', 0.0], ['Instagram', 0.0], ['Temple Run', 0.0]]

for app in apps_data:
    price = app[1]

    if price == 0.0:
        app.append('free')
    if price != 0.0:
        app.append('not free')

print(apps_data)

[['Call of Duty: Zombies', 5.0, 'not free'], ['Facebook', 0.0, 'free'], ['Instagram', 0.0, 'free'], ['Temple Run', 0.0, 'free']]


In the code above, we iterated through the apps_data list of lists and for each iteration we completed the following:

- We saved the price value to a variable named `price`.
- If `price == 0.0`, we appended the string '`free`' to the list app (app is the iteration variable).
- If `price != 0.0`, we appended the string '`not free`' to the list app.

For each iteration, the computer checked whether `price == 0.0` and `price != 0.0`. After we find out an app is `price == 0.0`, we don't need to check `price != 0.0` for the same app. We can avoid this redundancy by combining an `if` statement with an `else` clause:

In [32]:
apps_data = [['Call of Duty: Zombies', 5.0], ['Facebook', 0.0], ['Instagram', 0.0], ['Temple Run', 0.0]]

for app in apps_data:
    price = app[1]

    if price == 0.0:
        app.append('free')
    else:
        app.append('not free')

print(apps_data)

[['Call of Duty: Zombies', 5.0, 'not free'], ['Facebook', 0.0, 'free'], ['Instagram', 0.0, 'free'], ['Temple Run', 0.0, 'free']]


The code within the body of an **`else` clause executes *only if* the `if` statement that precedes it resolves to `False`**.

In [33]:
if False:
    print(1)
else:
    print(2)

2


In [34]:
if True:
    print(1)
else:
    print(2)

1


Here's an example that's like our example above:

In [35]:
price = 5

if price == 0:
    print('free')
else:
    print('not free')

not free


In our example above (the example at the beginning of this screen), the code within the body of the else clause executes only `if price == 0.0` evaluates to `False`. If `price == 0.0` is `True`, then the code `app.append('free')` executes and the computer moves forward without executing the `else` clause.

Note that we must combine an else clause with a preceding if statement. We can have an if statement without an else clause, but we can't have an else clause without a preceding if statement.

### Categorize our apps

Now, let us label each app as "free" or "not free" depending on its price.

In [36]:
opened_file = open('AppleStore.csv', encoding="utf-8")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

for app in apps_data[1:]:
    price = float(app[4])
    # Complete code from here


    if price == 0.0:
        app.append('free')
    else:
        app.append('not free')

apps_data[0].append('free_or_not')
print(apps_data[:6])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic', 'free_or_not'], ['281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1', 'not free'], ['281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1', 'free'], ['281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1', 'free'], ['282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1', 'free'], ['282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+',

Let's do some more specific labeling than just "free" and "not free." We want to label the apps using this convention:

![compare](figs/20-compare.png)

When an app is free, `price == 0.0` evaluates to `True` and `app.append('free')` executes. Then the computer continues to do other operations — it should check for the following:

- `price > 0 and price < 20`
- `price >= 20 and price < 50`
- `price >= 50`

We already know the three conditions above evaluate to `False` once we determine that `price == 0.0` is `True`. To stop the computer from doing redundant operations, we can use **`elif` clauses**:

In [37]:
apps_data = [['Facebook', 0.0], ['Notion', 14.99], ['Astropad Standard', 29.99], ['NAVIGON Europe', 74.99]]

for app in apps_data:
    price = app[1]

    if price == 0.0:
        app.append('free')
    elif price > 0.0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    elif price >= 50:
        app.append('very expensive')

print(apps_data)

[['Facebook', 0.0, 'free'], ['Notion', 14.99, 'affordable'], ['Astropad Standard', 29.99, 'expensive'], ['NAVIGON Europe', 74.99, 'very expensive']]


The code within the body of an `elif` clause executes only under the following circumstances:

- The preceding `if` statement (or the other preceding `elif` clauses) resolves to `False`; and
- The condition specified after the `elif` keyword evaluates to `True`.

If `price > 0.0 and price < 20` is `True`, then `app.append('affordable')` executes, and the remaining two elif clauses are disregarded. If `price >= 20 and price < 50` is `True`, then `app.append('expensive')` executes and the last `elif` clause is disregarded.

The `app_ratings` variable provided in the code below includes average user rating information for a few apps. Using `if` and `elif` clauses, we will append a statement that describes how each app rating compares to the overall average user rating of the entire `AppleStore.csv` dataset, which is `3.52`.

We will label each app as "below average", "roughly average", or "better than average" depending on its rating.

In [38]:
app_ratings = [['Facebook', 3.5], ['Notion', 4.0], ['Astropad Standard', 4.5], ['NAVIGON Europe', 3.5]]

for app in app_ratings:
    rating = app[1]
    
    if rating < 3.0:
        app.append('below average')
    elif rating >= 3.0 and rating < 4.0:
        app.append('roughly average')
    elif rating >= 4.0:
        app.append('better than average')
        
print(app_ratings)

[['Facebook', 3.5, 'roughly average'], ['Notion', 4.0, 'better than average'], ['Astropad Standard', 4.5, 'better than average'], ['NAVIGON Europe', 3.5, 'roughly average']]


It's possible to replace the final `elif` clause with an `else` clause. For our data, the output would be identical.

In [39]:
apps_data = [['Facebook', 0.0], ['Notion', 14.99], ['Astropad Standard', 29.99], ['NAVIGON Europe', 74.99]]

for app in apps_data:
    price = app[1]

    if price == 0.0:
        app.append('free')
    elif price > 0.0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    else:
        app.append('very expensive')

print(apps_data)

[['Facebook', 0.0, 'free'], ['Notion', 14.99, 'affordable'], ['Astropad Standard', 29.99, 'expensive'], ['NAVIGON Europe', 74.99, 'very expensive']]


Note, however, that the logic is different when we use an `else` clause. For `elif price >= 50`, `app.append('very expensive')` will execute only if the price is greater than or equal to 50. For `else`, there's no condition to meet (other than the previous `if` and `elif` clauses resolving to `False`), and `app.append('very expensive')` executes even if the price has a value of -5 or -100

In [40]:
apps_data = [['Facebook', 0.0], 
             ['Notion', 14.99], 
             ['Astropad Standard', -100], 
             ['NAVIGON Europe', -5]
            ]

for app in apps_data:
    price = app[1]

    if price == 0.0:
        app.append('free')
    elif price > 0.0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    else:
        app.append('very expensive')

print(apps_data)

[['Facebook', 0.0, 'free'], ['Notion', 14.99, 'affordable'], ['Astropad Standard', -100, 'very expensive'], ['NAVIGON Europe', -5, 'very expensive']]


Now, we will use our dataset to label each app as "free," "affordable," "expensive," or "very expensive."

In [41]:
opened_file = open('AppleStore.csv', encoding="utf-8")
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

for app in apps_data[1:]:
    price = float(app[4])

    if price == 0:
        app.append('free')
    elif price > 0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    elif price >= 50:
        app.append('very expensive')
        
apps_data[0].append('price_label')
print(apps_data[:6])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic', 'price_label'], ['281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1', 'affordable'], ['281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1', 'free'], ['281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1', 'free'], ['282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1', 'free'], ['282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+

## Summary

In this lesson, we learned to answer more advanced questions by combining conditional statements and clauses (`if, else, elif`) with logical operators (`and, or`) and comparison operators (`==, !=, >, >=, <, <=`).